File: ble-flet/bs2/wire.py
"""BS2 wire encode/decode — mirrors extension/src/bitstream2 framing."""

from __future__ import annotations

BS_PREFIX = b"BS "
BS_TYPE_REQ = 0x02
BS_TYPE_RES = 0x03
BS_TYPE_EVT_SENSOR = 0x04

BS_CMD_PING = 0x01
BS_CMD_SENSOR_CFG_GET = 0x10
BS_CMD_SENSOR_CFG_SET = 0x11
BS_CMD_BMI270_MODE_SET = 0x14
BS_CMD_BLE_POLICY_GET = 0x35
BS_CMD_BLE_POLICY_SET = 0x36

BLE_POLICY_BOOT_DEFAULT = 0x05
BLE_POLICY_FACTORY_STREAMING = 0x07


def crc16_ccitt(data: bytes, init: int = 0xFFFF) -> int:
    crc = init
    for b in data:
        crc ^= b << 8
        for _ in range(8):
            if crc & 0x8000:
                crc = ((crc << 1) ^ 0x1021) & 0xFFFF
            else:
                crc = (crc << 1) & 0xFFFF
    return crc


def encode_bs_req(req_id: int, cmd_id: int, body: bytes = b"", flags: int = 0) -> bytes:
    payload = bytes((req_id & 0xFF, (req_id >> 8) & 0xFF, cmd_id & 0xFF, flags & 0xFF)) + body
    plen = len(payload)
    header = BS_PREFIX + bytes((plen & 0xFF, (plen >> 8) & 0xFF, BS_TYPE_REQ))
    crc = crc16_ccitt(header[3:] + payload)
    return header + payload + bytes((crc & 0xFF, (crc >> 8) & 0xFF, 0x0D, 0x0A))


def try_parse_bs2_frame(wire: bytes) -> tuple[int, bytes] | None:
    if len(wire) < 14 or wire[0:3] != BS_PREFIX:
        return None
    plen = wire[3] | (wire[4] << 8)
    needed = 3 + 2 + 1 + plen + 2 + 2
    if len(wire) < needed or wire[needed - 2 : needed] != b"\r\n":
        return None
    crc_expected = wire[6 + plen] | (wire[7 + plen] << 8)
    crc_actual = crc16_ccitt(wire[3 : 6 + plen])
    if crc_actual != crc_expected:
        return None
    return wire[5], wire[6 : 6 + plen]


def try_parse_bs2_res(wire: bytes) -> tuple[int, int, int, bytes] | None:
    parsed = try_parse_bs2_frame(wire)
    if parsed is None:
        return None
    msg_type, payload = parsed
    if msg_type != BS_TYPE_RES or len(payload) < 4:
        return None
    req_id = payload[0] | (payload[1] << 8)
    cmd_id = payload[2]
    status = payload[3]
    body = payload[4:]
    return req_id, cmd_id, status, body


def try_parse_evt_sensor(wire: bytes) -> dict | None:
    parsed = try_parse_bs2_frame(wire)
    if parsed is None:
        return None
    msg_type, payload = parsed
    if msg_type != BS_TYPE_EVT_SENSOR or len(payload) < 10:
        return None
    sensor_id = payload[0]
    mask = payload[1]
    counter = int.from_bytes(payload[2:6], "little")
    device_ms = int.from_bytes(payload[6:10], "little")
    values_bytes = payload[10:]
    return {
        "sensor_id": sensor_id,
        "mask": mask,
        "counter": counter,
        "device_ms": device_ms,
        "values_bytes": values_bytes,
    }

File: ble-react/packages/tbs-core/src/wire.ts
/** BS2 framed protocol — wire encode/decode */

export const BS_PREFIX = new Uint8Array([0x42, 0x53, 0x20]); // "BS "
export const BS_TYPE_REQ = 0x02;
export const BS_TYPE_RES = 0x03;
export const BS_TYPE_EVT_SENSOR = 0x04;

export function crc16Ccitt(data: Uint8Array, init = 0xffff): number {
  let crc = init;
  for (let i = 0; i < data.byteLength; i++) {
    crc ^= data[i]! << 8;
    for (let b = 0; b < 8; b++) {
      if (crc & 0x8000) {
        crc = ((crc << 1) ^ 0x1021) & 0xffff;
      } else {
        crc = (crc << 1) & 0xffff;
      }
    }
  }
  return crc;
}

export function encodeBsReq(reqId: number, cmdId: number, body: Uint8Array = new Uint8Array(), flags = 0): Uint8Array {
  const payload = new Uint8Array(4 + body.byteLength);
  payload[0] = reqId & 0xff;
  payload[1] = (reqId >> 8) & 0xff;
  payload[2] = cmdId & 0xff;
  payload[3] = flags & 0xff;
  payload.set(body, 4);

  const plen = payload.byteLength;
  const out = new Uint8Array(3 + 2 + 1 + plen + 2 + 2);
  out.set(BS_PREFIX, 0);
  out[3] = plen & 0xff;
  out[4] = (plen >> 8) & 0xff;
  out[5] = BS_TYPE_REQ;
  out.set(payload, 6);

  const crcRegion = out.subarray(3, 6 + plen);
  const crc = crc16Ccitt(crcRegion);
  out[6 + plen] = crc & 0xff;
  out[7 + plen] = (crc >> 8) & 0xff;
  out[8 + plen] = 0x0d;
  out[9 + plen] = 0x0a;
  return out;
}

export function tryParseBs2Frame(wire: Uint8Array): { msgType: number; payload: Uint8Array } | null {
  if (wire.byteLength < 14) return null;
  if (wire[0] !== 0x42 || wire[1] !== 0x53 || wire[2] !== 0x20) return null;
  const plen = wire[3]! | (wire[4]! << 8);
  const needed = 3 + 2 + 1 + plen + 2 + 2;
  if (wire.byteLength < needed) return null;
  if (wire[needed - 2] !== 0x0d || wire[needed - 1] !== 0x0a) return null;
  const crcExpected = wire[6 + plen]! | (wire[7 + plen]! << 8);
  const crcActual = crc16Ccitt(wire.subarray(3, 6 + plen));
  if (crcActual !== crcExpected) return null;
  return { msgType: wire[5]!, payload: wire.subarray(6, 6 + plen) };
}

export function tryParseBs2Res(
  wire: Uint8Array,
): { reqId: number; cmdId: number; status: number; body: Uint8Array } | null {
  const parsed = tryParseBs2Frame(wire);
  if (!parsed || parsed.msgType !== BS_TYPE_RES || parsed.payload.byteLength < 4) return null;
  const p = parsed.payload;
  return {
    reqId: p[0]! | (p[1]! << 8),
    cmdId: p[2]!,
    status: p[3]!,
    body: p.subarray(4),
  };
}

export type EvtSensorRaw = {
  sensorId: number;
  mask: number;
  counter: number;
  deviceMs: number;
  values: Uint8Array;
};

export function tryParseEvtSensor(wire: Uint8Array): EvtSensorRaw | null {
  const parsed = tryParseBs2Frame(wire);
  if (!parsed || parsed.msgType !== BS_TYPE_EVT_SENSOR || parsed.payload.byteLength < 10) return null;
  const p = parsed.payload;
  const view = new DataView(p.buffer, p.byteOffset, p.byteLength);
  return {
    sensorId: p[0]!,
    mask: p[1]!,
    counter: view.getUint32(2, true),
    deviceMs: view.getUint32(6, true),
    values: p.subarray(10),
  };
}

export function looksLikeBsPrefix(data: Uint8Array): boolean {
  return data.byteLength >= 3 && data[0] === 0x42 && data[1] === 0x53 && data[2] === 0x20;
}
